### DISCUSSION TITLE
Serving multiple teams on different release cadences

### DISCUSSION THREAD

**Discussion by lorenzodejong** [nrwl/nx#17832] [General]
Hey all,

I have a question regarding the setup of a monorepo which is suited for different teams which are on different release cadences.

So let's say we have the following setup:

Team A:
- Owns application A, releasing every 2 weeks
- Owns application B, releasing every day

Team B:
- Owns application C, releasing every 2 weeks
- Owns application D, releasing every 2 months

All applications depend on some shared libraries, causing them to become "affected" before the potential release moment. So deriving the release strategy from `nx affected` would not work in this scenario, unless we strictly enforce feature flags for every application.

A potential option would be to enforce a convention-based release strategy, for example by creating an application scoped release branch (`release/app-a/1.0.0`) on which we can base which application should be deployed. However i'm not a big fan of these convention-based approaches.

My question is if it would be advisable to setup our monorepo in this way, and if so what would be a potential solution for implementing the CI/CD setup with this strategy?

**Comment by alacret** [comment]
@lorenzodejong have you found any info here?

**Comment by lorenzodejong** [comment]
@alacret basically together with my team we came to the conclusion that there's three general approaches:

### Splitting up apps with different release cadences

This is the simplest approach of them all. Split up apps which don't share the same release cadence into a separate monorepo.

`+` Full separation of the CI/CD
`-` Harder to share packages between monorepos

### Convention-based releases

As mentioned in my initial post, release an app using an explicit trigger from a chosen convention. Potential triggers you can use:

- Release branches: `release/app-a/1.0.0`
- Tags: `app-a/1.0.0`
- Variable from a release pipeline

Based on the app name variable from the chosen trigger you can determine for which app to run the Nx command to initiate the deployment.

`+` Control over which apps releases at what time
`-` Specific conventions which need to be aligned with the team
`-` Not a fully integrated CI/CD workflow, since a manual release is still required

### Trunk-based development (with feature flags)

This approach requires the highest level of seniority in your team, since you're always treating the trunk branch as the state which can be integrated to production. Branches should ideally be short-lived, features can be switched off in production by using feature flags.

https://trunkbaseddevelopment.com/

`+` Most linear workflow, as the state on the trunk represent the state on production
`-` Difficult to get right, requires a high level of test coverage and seniority for correctly managing feature flags

### Conclusion

There's no clear right or wrong for the solution you choose. This choice should be based on level of seniority in your team, alignment of the release cadences and other potential external dependencies of your apps (which are on a specific release cadence).

**Reply by alacret** [reply]
Yes, this is kind of what we reach out to
